### PR TITLE
Correctly escape paths in file names

### LINKED ISSUE #3639: 🐛 BUG: Building a site using React components with escaped spaces in path fails
### What version of `astro` are you using?

1.0.0-beta.47

### Are you using an SSR adapter? If so, which one?

Node

### What package manager are you using?

yarn

### What operating system are you using?

Windows

### Describe the Bug

When building an Astro site using the Node SSR adapter, as well with a React component placed in a folder with a space in it's name will fail.

Vite is unable to load the component during build time, thus failing the build.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-mhnkxf

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by bluwy**
## Changes

Fix #3639

Exotic file names that includes ` `, `^`, `🚀`, `(`, `{` are now handled and encoded properly.

`%` is not supported for now as I found Vite had trouble processing it too.

Note: Use emojis in file names, the HTML must have `charset="utf-8"` to work.

## Testing

Enable tests added from https://github.com/withastro/astro/pull/4263

## Docs

N/A

### CODE REVIEW

File: packages/astro/src/runtime/server/hydration.ts [hunk 1]
```diff
@@ -140,12 +140,12 @@ export async function generateHydrateScript(
 	}
 
 	// Add component url
-	island.props['component-url'] = await result.resolve(componentUrl);
+	island.props['component-url'] = await result.resolve(decodeURI(componentUrl));
```
Comment (bluwy) [inline packages/astro/src/runtime/server/hydration.ts:143]: `componentUrl` is retrieved as-is from the attribute, which is encoded. In code, we want to deal with the decoded version instead.

File: packages/astro/src/jsx/babel.ts [hunk 2]
```diff
@@ -299,11 +297,8 @@ export default function astroJSX(): PluginObj {
...
 					let resolvedPath: string;
 					if (meta.path.startsWith('.')) {
-						const fileURL = pathToFileURL(state.filename!);
-						resolvedPath = `/@fs${new URL(meta.path, fileURL).pathname}`;
-						if (resolvedPath.endsWith('.jsx')) {
-							resolvedPath = resolvedPath.slice(0, -4);
-						}
+						resolvedPath = resolve(meta.path, state.filename!);
+						resolvedPath = resolveClientDevPath(resolvedPath);
 					} else {
```
Comment (bluwy) [inline packages/astro/src/jsx/babel.ts:302]: This code runs in node, but uses the URL api which auto encodes paths by default. Switch to using `path.resolve` and handle the `.jsx` condition with `resolveClientDevPath`.

I also removed the `/@fs` part as tracing through it's usage, this isn't needed. This path is only used to pass as Vite entry, which can read from the filesystem directly as absolute path.

File: packages/astro/src/core/build/vite-plugin-analyzer.ts [hunk 3]
```diff
@@ -82,7 +82,7 @@ export function vitePluginAnalyzer(internals: BuildInternals): VitePlugin {
 				const astro = info.meta.astro as AstroPluginMetadata['astro'];
 
 				for (const c of astro.hydratedComponents) {
-					const rid = c.resolvedPath ? resolveClientDevPath(c.resolvedPath) : c.specifier;
+					const rid = c.resolvedPath ?? c.specifier;
```
Comment (bluwy) [inline packages/astro/src/core/build/vite-plugin-analyzer.ts:85]: `resolvedPath` has already called `resolveClientDevPath` earlier, so we remove here.

**Review by natemoo-re — COMMENTED** [review]
This looks great! E2E tests are failing though, any insights?

**Comment by Princesseuh** [comment]
It took three Astro core members, but we finally got this fixed 💪 Awesome work

**Comment by bluwy** [comment]
Wow looks like that e2e test alone is quite a rabbithole (again 🥲 ), I can reproduce it locally too. I'll try to look into it tomorrow, marking as draft for now.

**Comment by hippotastic** [comment]
Sorry for building a set of rabbit holes! 😆 But I'm really happy that we're making progress here in getting the tests green! Great work!

File: packages/astro/src/jsx/babel.ts (hunk 2 above)
Comment (bluwy) [reply packages/astro/src/jsx/babel.ts]: I reverted this change. I thought `resolvedPath` meant the filesystem path, but it seems to be treated as a URL that starts with `/@fs` everywhere.

**Comment by bluwy** [comment]
This should be good now. Luckily the CI errors were actually real problems. I toned down some refactors earlier so it should fix specifically the escaping issues.

**Comment by hippotastic** [comment]
Yay, I'm so excited! Finally a working solution to this unexpectedly hard problem. Glad we have a Vite wizard on board now. Thank you so much, @bluwy!

**Review by hippotastic — APPROVED** [review]
I'm only the test author and not a core member, but that LGTM!

**Review by natemoo-re — APPROVED** [review]
Incredible work!

**Pull request merged**